### ISSUE TITLE
acorn should be in dependencies (breaking Yarn PnP builds)

### ISSUE THREAD

**Issue by tarikjn** [terser/terser#1076]
terser is currently breaking builds using Yarn PnP as a package manager. It appears that the issue would be `acorn` being incorrectly listed in `devDependencies` instead of `dependencies` in terser's `package.json`: https://github.com/terser/terser/blob/ecb869e75637e3e3db1c6d34b2b2f8bdb68f01c4/package.json#L52
[](url)
```
ERROR in ./.yarn/cache/terser-npm-5.9.0-16d6a713ce-11c1246b19.zip/node_modules/terser/dist/bundle.min.js 27827:31-53
Module not found: Error: Can't resolve 'acorn' in '/Users/tarik/Workspace/test/.yarn/cache/terser-npm-5.9.0-16d6a713ce-11c1246b19.zip/node_modules/terser/dist'
```

For more details, see the original issue reported on another package that depends on parser: https://github.com/terser/html-minifier-terser/issues/73

**Comment by Conduitry** [comment]
Terser only uses Acorn if it is instructed to do so - https://github.com/terser/terser#use-acorn-for-parsing - If you're calling Terser in this way, Acorn effectively becomes a peer dependency, but there's no production dependency on Acorn otherwise.

**Comment by DanielRuf** [comment]
> Is Yarn trying to aggressively follow dependencies wherever it sees a `require()`?

That is how Yarn PnP works. See https://yarnpkg.com/features/pnp/#pnp-loose-mode. So unless you use the `loose` mode this is the right behavior of yarn v2 (also called berry). And to be compatible with Yarn PnP Terser in fact has to list this dependency (even if it may be unused in most cases).

**Comment by fabiosantoscode** [comment]
Makes sense, let's do it

**Comment by Conduitry** [comment]
Does Yarn 2 support the new-ish concept of optional peer dependencies that npm does? Making everyone using Terser download Acorn (that they almost certainly won't use) for the sake of Yarn sound like the wrong solution.

**Comment by fabiosantoscode** [comment]
Yarn is not the first (and won't be the last) thing that's interested in uncritically following all require() or import(). I think there's great interest in allowing for static analysis of code.

**Comment by DanielRuf** [comment]
> optional peer dependencies

Not that I knew, see https://classic.yarnpkg.com/en/docs/dependency-types/

There is not much "optional" and there were good reasons for the PnP API.

**Comment by Conduitry** [comment]
Welp if this decision has been made, I guess it'll just be another thing to add to my pnpmfile to override the `pkg.dependencies` of Terser.

**Comment by Conduitry** [comment]
I just found https://github.com/yarnpkg/rfcs/blob/master/accepted/0000-optional-peer-dependencies.md - I would strongly advise at least trying this first. This is the syntax that was later adopted by npm, and I would be surprised if it weren't first implemented in Yarn, since that's where the RFC originated.

**Comment by fabiosantoscode** [comment]
Which raises the question, is acorn a peer dependency of Terser?

Most importantly, what is a peer dependency, and is it safe to use? The meaning and mechanics of it has changed multiple times in the past years at least in npm.

**Comment by lukeed** [comment]
This works in all npm & pnpm versions:

```js
"peerDependencies": { 
  "acorn": "*" // can specify ">=8.5" if wanted
},
"peerDependenciesMeta": {
  "acorn": {
    "optional": true
  }
}
```

IIRC, in some older npm versions, `peerDependenciesMeta` is ignored and in that worse-case scenario (from @Conduitry's stance), anything in `peerDependencies` is automatically downloaded. Prehistoric npm versions didn't download `peerDependencies` but would link correctly when user downloads the peer.

The Yarn RFC that @Conduitry linked is the same format above, so we should trust/assume that berry will support that format too – especially since that's already the majority support.

TLDR: If Terser is to make any changes, it should be the snippet above.

**Comment by jridgewell** [comment]
[esbuild](https://github.com/evanw/esbuild/releases) just transitioned to optional dependencies for its install process, and it works fine for pnp mode (they had unrelated difficulties that we won't have). I'd recommend moving an optional dependency so that npm doesn't tell us "unmet peer dependency" every time terser is installed.

**Comment by fabiosantoscode** [comment]
I'm on it atm, just noticed some tests weren't running. We ran out of credits on Travis again :D 

Their business model changed, so we need to transition to GH actions eventually.

**Comment by fabiosantoscode** [comment]
Boom!

**Comment by fabiosantoscode** [comment]
Thanks everyone, for bringing the problem up and debating the solution :)

**Comment by dietergeerts** [comment]
It's sad to see this change as a minor one, as it breaks builds with pnpm, which checks peer dependencies. For example, the following package was using acorn v7: https://github.com/storybookjs/storybook/blob/v6.3.12/addons/docs/package.json, and it now breaks with pnpm due to this change incorrectly stated as breaking change.

I would like to see this corrected to not break builds.

**Comment by dietergeerts** [comment]
#:storybookjs#16776

**Comment by dietergeerts** [comment]
#storybookjs/storybook#16776

**Comment by fabiosantoscode** [comment]
Hello @dietergeerts 

Thanks for another reminder of why peer dependencies are garbage.

While in practice the current solution may allow acorn to be optional, the semantics do cause unnecessary dependency hell.

I don't want to widen the acceptable version of Acorn because that just seems like digging deeper into this kind of issues, and will require me to keep up with Acorn versions as they come out.

I'll be converting this to an actual dependency, which will use up more space but at least will let Terser work with all package managers.

**Comment by dietergeerts** [comment]
> Thanks for another reminder of why peer dependencies are garbage.

I would not say that, though I understand your reasoning.

From what I see a lot, peer dependencies are used wrong most of the time, and that's what's causing problems. If peer dependencies where to be used correctly, there wouldn't be any problems in any situation. I'll give some examples about what I mean, and how I see it:

In lots of cases I see peer dependencies used for dependencies that, if they are available, they will be used by the package, so the "optional". But this kind of dependencies is just not a peer dependencies. The package don't need it to work, and setting it "optional" indicates that part of the system is pluggable. If this is the case, then creating a new package that can be used together with those dependencies is a far better and correct solution, as then the consumer actually states they want to use that "optional" dependency. Lots of packages check if such "optional" dependency is there and then us it, but what if that dep is there due to some other package, and it was not meant to be used in this package? See why automatic dependency loading and all other auto stuff is actually not a good idea for all situations? Far better is to set things explicitly.

A peer dependency is actually very simple. Does your package depend on something which it can't control, or may not control, then it's a peer dependency, like when your package is a plugin for some other package, or when you need a global instance (singleton) to be there from another package, or when you explicitly want some instance of something given to use, but that can be used by other things in the consumer package, so you can't control it.

Now I don't know for what acorn is used in this package, and I believe there is a better solution like e.g. described above, but I also am aware that these solutions, though better and working in all situations, are asking more time and effort to implement and maintain.

Just wanted to share my thoughts a bit on this topic, as I'm seeing more and more "misuse" of dependencies, either adding "optional" peer dependencies, which is something that just isn't correct, either by adding dependencies that should be peer dependencies, like a lot of packages in the react ecosystem do, just for the sake of having to add less deps in the consuming packages, which off course result in too much problems.
